docs/source/traits_user_manual/internals.rst

```diff
@@ -0,0 +1,190 @@
+itself, and for those using Traits who would like a better understanding of
+the Traits internals.
+
+
+Fundamentals
+------------
+
+The runtime behavior of Traits is governed by two key classes: |CTrait| and
+|HasTraits|. Each of these classes inherits from a superclass implemented in
+the |ctraits| extension module: |CTrait| inherits from |cTrait|, while
+|HasTraits| inherits from |CHasTraits|.
+
+The |CHasTraits| and |cTrait| base classes are not intended for direct use and
+do not provide a complete, coherent API: that API is provided by their
+respective subclasses |HasTraits| and |CTrait|. The existence and precise
+semantics of the base classes should be regarded as implementation details.
+In what follows, we'll frequently refer to behavior of |HasTraits| and |CTrait|
+even when the behavior we're referring to is actually implemented by
+|CHasTraits| or |cTrait|.
+
+We'll examine both of these classes, along with some of their hidden state, in
+more detail below.
+
+
+Key class: |HasTraits|
+~~~~~~~~~~~~~~~~~~~~~~
+
+The primary purpose of the |HasTraits| class is to override Python's normal
+rules for attribute access. It does this by overriding the
+``__getattribute__``, ``__setattr__`` and ``__delattr__`` special methods. At C
+level, this is done through providing ``tp_getattro`` and ``tp_setattro``
+"slots" to the |CHasTraits| ``PyTypeObject``: ``tp_getattro`` controls
+attribute retrieval operations, while ``tp_setattro`` is called for both
+attribute set and attribute deletion operations.
+
+To support trait get and set operations, a |HasTraits| object (that is, in
+normal use, an instance of a user-defined subclass of |HasTraits|) has two key
+pieces of state: a dictionary of **class traits**, and a second dictionary of
+**instance traits**. Each dictionary is a mapping from trait attribute names to
+|CTrait| objects. A |CTrait| object encapsulates the rules for getting and
+setting the corresponding attribute, as well as providing an attachment point
+for trait notifications.
+
+In addition to these two dictionaries, a |HasTraits| object has a ``__dict__``
+that stores attribute values for that object in the normal way.
+
+Analogously to class variables and instance variables in a normal Python class,
+the class traits dictionary for a given |HasTraits| subclass is shared between
+instances of that class, while the instance traits dictionary is specific to a
+particular |HasTraits| instance. (As an internal optimization, the instance
+traits dictionary is created on demand when first needed, while the class
+traits dictionary is always present for each instance.)
+
+For introspection and debugging purposes, both of these dictionaries can
+be retrieved directly, using the |_instance_traits| and |_class_traits|
+methods. However, it's not recommended to use either of these methods in
+production code, and you should be especially careful when modifying either
+of these dictionaries.
```

Comment:
Just wondering if it would be useful to mention that the class traits can be accessed via a public class method `class_traits` on a subclass of `HasTraits`.

Reply:
I'm a bit nervous about mentioning the `class_traits` method here since it doesn't actually seem to do what we think it should: I'm not confident that it actually *does* reliably retrieve the class traits.

Reply:
Thanks!
I don't quite understand how #694 is related to class_traits... or did you mean #674?  Indeed #674 kills the idea of mentioning `class_traits` in this document and I forgot about it. Thanks for reminding me!

Reply:
Sorry; yes, I did mean #674.